src/d3a/models/area/markets.py

```diff
@@ -15,84 +15,158 @@
+        self.indexed_future_markets = {m.id: m for m in self.all_future_spot_markets}
 
-    @property
-    def all_future_spot_markets(self):
-        return list(self.markets.values())
+    @staticmethod
+    def _is_it_time_to_delete_past_settlement_market(current_time_slot: DateTime,
+                                                     time_slot: DateTime) -> bool:
+        """Check if the past settlement market for time_slot is ready to be deleted."""
+        return (time_slot < current_time_slot.subtract(
+                    hours=ConstSettings.GeneralSettings.MAX_AGE_SETTLEMENT_MARKET_HOURS,
+                    minutes=GlobalConfig.slot_length.total_minutes()))
 
-    def rotate_markets(self, current_time):
-        # Move old and current markets & balancing_markets to
-        # `past_markets` & past_balancing_markets. We use `list()` here to get a copy since we
-        # modify the market list in-place
-        self._market_rotation(current_time=current_time, markets=self.markets,
-                              past_markets=self.past_markets)
-        if self.balancing_markets is not None:
-            self._market_rotation(current_time=current_time, markets=self.balancing_markets,
-                                  past_markets=self.past_balancing_markets)
-        self._indexed_future_markets = {
-            m.id: m for m in self.all_future_spot_markets
-        }
-
-    def _market_rotation(self, current_time, markets, past_markets):
-        for timeframe in list(markets.keys()):
-            if timeframe < current_time:
-                market = markets.pop(timeframe)
+    @staticmethod
+    def _is_it_time_to_delete_past_market(current_time_slot: DateTime,
+                                          time_slot: DateTime) -> bool:
+        """Check if the past (balancing-)market for time_slot is ready to be deleted."""
+
+        if constants.RETAIN_PAST_MARKET_STRATEGIES_STATE:
+            return False
+
+        if ConstSettings.GeneralSettings.ENABLE_SETTLEMENT_MARKETS:
+            return (time_slot < current_time_slot.subtract(
+                hours=ConstSettings.GeneralSettings.MAX_AGE_SETTLEMENT_MARKET_HOURS))
+        else:
+            return time_slot < current_time_slot.subtract(
+                minutes=GlobalConfig.slot_length.total_minutes())
+
+    def rotate_markets(self, current_time: DateTime) -> None:
+        """Deal with market rotation of different types."""
+        self._move_markets_to_past(current_time)
+        self._delete_past_markets(current_time)
+
+        if ConstSettings.BalancingSettings.ENABLE_BALANCING_MARKET:
+            self._move_balancing_markets_to_past_balancing(current_time)
+            self._delete_past_balancing_markets(current_time)
+
+        if ConstSettings.GeneralSettings.ENABLE_SETTLEMENT_MARKETS:
+            self._move_settlement_markets_to_past_settlement(current_time)
+            self._delete_past_settlement_markets(current_time)
```

Comment:
Nit: this seems to be a good place for refactoring. All these blocks do two operations: "move to past" and "delete" markets. So this looks like a good place to implement an interface with 3 different implementations. E.g.: `MarketsRotator`, `BalancingMarketsRotator`, `SettlementMarketsRotator` etc.:
```python
# In the init
self.markets_rotator = MarketsRotatorMapper.get_market_rotator()
# And later on
self.markets_rotator.rotate()
```
Just an idea.

Reply:
I did that here:
https://github.com/gridsingularity/d3a/pull/1213/commits/121e754b98467c6bde7f54d67df6c27bbc5526ca
Unfortunately, also the slot markets depend on the settlement market setting (need to keep more past markets in memory in such a case), so the MarketRotator is not free of settlementmarket logic. 
But still I like the change. Let me know if you agree, Thanks!